components/powersensor/renogy/renogy.go

```diff
@@ -0,0 +1,277 @@
+// Package renogy implements the renogy charge controller sensor
+// Tested with renogy wanderer model
+// Wanderer Manual: https://www.renogy.com/content/RNG-CTRL-WND30-LI/WND30-LI-Manual.pdf
+// LCD Wanderer Manual: https://ca.renogy.com/content/manual/RNG-CTRL-WND10-Manual.pdf
+package renogy
+
+import (
+	"context"
+	"encoding/binary"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/goburrow/modbus"
+
+	"go.viam.com/rdk/components/powersensor"
+	"go.viam.com/rdk/resource"
+)
+
+var (
+	globalMu sync.Mutex
+	model    = resource.DefaultModelFamily.WithModel("renogy")
+	readings map[string]interface{}
+)
+
+const (
+	// defaults assume the device is connected via UART serial.
+	pathDefault     = "/dev/serial0"
+	baudDefault     = 9600
+	modbusIDDefault = 1
+
+	solarVoltReg             = 263
+	solarAmpReg              = 264
+	solarWattReg             = 265
+	loadVoltReg              = 260
+	loadAmpReg               = 261
+	loadWattReg              = 262
+	battVoltReg              = 257
+	battChargePctReg         = 256
+	controllerDegCReg        = 259
+	maxSolarTodayWattReg     = 271
+	minSolarTodayWattReg     = 272
+	maxBattTodayVoltReg      = 268
+	minBattTodayVoltReg      = 267
+	maxSolarTodayAmpReg      = 269
+	minSolarTodayAmpReg      = 270
+	chargeTodayWattHrsReg    = 273
+	dischargeTodayWattHrsReg = 274
+	chargeTodayAmpHrsReg     = 275
+	dischargeTodayAmpHrsReg  = 276
+	totalBattOverChargesReg  = 278
+	totalBattFullChargesReg  = 279
+)
+
+// Config is used for converting config attributes.
+type Config struct {
+	resource.TriviallyValidateConfig
+	Path     string `json:"serial_path,omitempty"`
+	Baud     int    `json:"serial_baud_rate,omitempty"`
+	ModbusID byte   `json:"modbus_id,omitempty"`
+}
+
+func init() {
+	resource.RegisterComponent(
+		powersensor.API,
+		model,
+		resource.Registration[powersensor.PowerSensor, *Config]{
+			Constructor: newRenogy,
+		})
+}
+
+func newRenogy(_ context.Context, _ resource.Dependencies, conf resource.Config, logger golog.Logger) (powersensor.PowerSensor, error) {
+	newConf, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	if newConf.Path == "" {
+		newConf.Path = pathDefault
+	}
+	if newConf.Baud == 0 {
+		newConf.Baud = baudDefault
+	}
+	if newConf.ModbusID == 0 {
+		newConf.ModbusID = modbusIDDefault
+	}
+
+	return &Renogy{
+		Named:    conf.ResourceName().AsNamed(),
+		path:     newConf.Path,
+		baud:     newConf.Baud,
+		modbusID: newConf.ModbusID,
+	}, nil
+}
+
+// Renogy is a serial charge controller.
+type Renogy struct {
+	resource.Named
+	resource.AlwaysRebuild
+	resource.TriviallyCloseable
```

Comment:
should we implement a close where we close the handler? Fine with leaving it as is as it doesn't stop the server from hanging or the renogy from re configuring.